File: backend/ml/utils/textNormalization.js
const SLANG_REPLACEMENTS = [
  [/\bim\b/g, 'i am'],
  [/\bive\b/g, 'i have'],
  [/\bdont\b/g, 'do not'],
  [/\bcant\b/g, 'cannot'],
  [/\bwont\b/g, 'will not'],
  [/\bdoesnt\b/g, 'does not'],
  [/\bdidnt\b/g, 'did not'],
  [/\bshouldnt\b/g, 'should not'],
  [/\bwouldnt\b/g, 'would not'],
  [/\bthere's\b/g, 'there is'],
  [/\btheres\b/g, 'there is'],
  [/\bpls\b/g, 'please'],
  [/\bplz\b/g, 'please'],
  [/\bu\b/g, 'you'],
  [/\bur\b/g, 'your'],
  [/\bgonna\b/g, 'going to'],
  [/\bwanna\b/g, 'want to'],
  [/\bkinda\b/g, 'kind of'],
  [/\bsorta\b/g, 'sort of'],
  [/\byaar\b/g, 'friend'],
  [/\bghabrahat\b/g, 'anxiety'],
  [/\bbechain\b/g, 'restless'],
  [/\budas\b/g, 'sad'],
  [/\budasi\b/g, 'sadness'],
  [/\bumeed\b/g, 'hope'],
  [/\bmehfooz\b/g, 'safe'],
  [/\bbohat\b/g, 'very'],
  [/\bbhot\b/g, 'very'],
  [/\bthora\b/g, 'a little'],
  [/\baaj\b/g, 'today'],
  [/\bkal\b/g, 'tomorrow'],
];

function normalizeText(text) {
  if (typeof text !== 'string') return '';

  let normalized = text.normalize('NFKC').toLowerCase().trim();
  normalized = normalized.replace(/[“”]/g, '"').replace(/[‘’]/g, "'");
  normalized = normalized.replace(/[^\p{L}\p{N}\s'’/-]+/gu, ' ');

  for (const [pattern, replacement] of SLANG_REPLACEMENTS) {
    normalized = normalized.replace(pattern, replacement);
  }

  normalized = normalized.replace(/\s+/g, ' ').trim();
  return normalized;
}

function stripPunctuation(text) {
  return normalizeText(text).replace(/[\/'-]/g, ' ').replace(/\s+/g, ' ').trim();
}

function makeCasualVariant(text) {
  return normalizeText(text)
    .replace(/\byou are\b/g, 'youre')
    .replace(/\bdo not\b/g, 'dont')
    .replace(/\bcannot\b/g, 'cant')
    .replace(/\bi am\b/g, "i'm")
    .replace(/\bi have\b/g, "i've");
}

function makeRomanUrduVariant(text) {
  return normalizeText(text)
    .replace(/\banxiety\b/g, 'ghabrahat')
    .replace(/\bsadness\b/g, 'udas')
    .replace(/\bsad\b/g, 'udas')
    .replace(/\bhope\b/g, 'umeed')
    .replace(/\bsafe\b/g, 'mehfooz')
    .replace(/\brestless\b/g, 'bechain')
    .replace(/\bvery\b/g, 'bohat');
}

function makeNoisyVariant(text) {
  return stripPunctuation(text)
    .replace(/\bvery\b/g, 'vry')
    .replace(/\bplease\b/g, 'pls')
    .replace(/\bfriend\b/g, 'yaar')
    .replace(/\btoday\b/g, 'aaj');
}

module.exports = {
  normalizeText,
  stripPunctuation,
  makeCasualVariant,
  makeRomanUrduVariant,
  makeNoisyVariant,
};

File: backend/ml_service/utils/text_normalization.py
"""Shared text preprocessing for the Python ML service.

This mirrors the JavaScript normalization layer closely so the JS fallback and
Python microservice see near-identical text before vectorization.
"""

from __future__ import annotations

import re
import unicodedata

SLANG_REPLACEMENTS = [
    (r"\bim\b", "i am"),
    (r"\bive\b", "i have"),
    (r"\bdont\b", "do not"),
    (r"\bcant\b", "cannot"),
    (r"\bwont\b", "will not"),
    (r"\bdoesnt\b", "does not"),
    (r"\bdidnt\b", "did not"),
    (r"\bshouldnt\b", "should not"),
    (r"\bwouldnt\b", "would not"),
    (r"\bthere's\b", "there is"),
    (r"\btheres\b", "there is"),
    (r"\bpls\b", "please"),
    (r"\bplz\b", "please"),
    (r"\bu\b", "you"),
    (r"\bur\b", "your"),
    (r"\bgonna\b", "going to"),
    (r"\bwanna\b", "want to"),
    (r"\bkinda\b", "kind of"),
    (r"\bsorta\b", "sort of"),
    (r"\byaar\b", "friend"),
    (r"\bghabrahat\b", "anxiety"),
    (r"\bbechain\b", "restless"),
    (r"\budas\b", "sad"),
    (r"\budasi\b", "sadness"),
    (r"\bumeed\b", "hope"),
    (r"\bmehfooz\b", "safe"),
    (r"\bbohat\b", "very"),
    (r"\bbhot\b", "very"),
    (r"\bthora\b", "a little"),
    (r"\baaj\b", "today"),
    (r"\bkal\b", "tomorrow"),
]

COMMON_TYPOS = [
    (r"\bdefinately\b", "definitely"),
    (r"\brecieve\b", "receive"),
    (r"\bseperate\b", "separate"),
    (r"\boccurence\b", "occurrence"),
]


def _normalize_repeated_characters(text: str) -> str:
    return re.sub(r"([a-z])\1{2,}", r"\1\1", text)


def normalize_text(text: str) -> str:
    if not isinstance(text, str):
        return ""

    normalized = unicodedata.normalize("NFKC", text).lower().strip()
    normalized = normalized.replace("“", '"').replace("”", '"').replace("‘", "'").replace("’", "'")
    normalized = re.sub(r"[^\w\s'/-]+", " ", normalized, flags=re.UNICODE)

    for pattern, replacement in COMMON_TYPOS + SLANG_REPLACEMENTS:
        normalized = re.sub(pattern, replacement, normalized)

    normalized = _normalize_repeated_characters(normalized)
    normalized = re.sub(r"\s+", " ", normalized).strip()
    return normalized


def strip_punctuation(text: str) -> str:
    return re.sub(r"\s+", " ", re.sub(r"[\/'-]", " ", normalize_text(text))).strip()


def make_casual_variant(text: str) -> str:
    return (
        normalize_text(text)
        .replace("you are", "youre")
        .replace("do not", "dont")
        .replace("cannot", "cant")
        .replace("i am", "i'm")
        .replace("i have", "i've")
    )


def make_roman_urdu_variant(text: str) -> str:
    return (
        normalize_text(text)
        .replace("anxiety", "ghabrahat")
        .replace("sadness", "udas")
        .replace("sad", "udas")
        .replace("hope", "umeed")
        .replace("safe", "mehfooz")
        .replace("restless", "bechain")
        .replace("very", "bohat")
    )


def make_noisy_variant(text: str) -> str:
    return (
        strip_punctuation(text)
        .replace("very", "vry")
        .replace("please", "pls")
        .replace("friend", "yaar")
        .replace("today", "aaj")
    )
